### ISSUE TITLE
Spaceport reporting fails after a while

### ISSUE THREAD

**Issue by hrkfdn** [apollographql/router#1903]
**Describe the bug**
After several hours/days of uptime, the router seems to be incapable of sending reports to Spaceport/Apollo Studio:
```
{"timestamp":"2022-09-28T13:13:42.666229Z","level":"WARN","fields":{"message":"failed to submit stats to spaceport: status: Unavailable, message: \"error trying to connect: tcp connect error: Connection refused (os error 111)\", details: [], metadata: MetadataMap { headers: {} }"},"target":"apollo_router::plugins::telemetry::apollo_exporter"}
```

Restarting the router seems to fix it until it fails again after a period of time. The error (Connection refused) seems to suggest that the target host cannot be reached, but I can curl the reporting URL (https://usage-reporting.api.apollographql.com/api/ingress/traces -- is the one that's used?) without problems.

This is particularly problematic because it also prevents sensible breaking change checks for schema updates. We did not experience this problem with the Apollo Gateway running in the same setup.

**To Reproduce**
1. Setup router with Apollo Studio metrics
2. Wait for several hours/days
3. See error

**Desktop (please complete the following information):**
 - Apollo Router 1.0.0

**Comment by garypen** [comment]
Transient errors may be encountered as the router is operating. The router is designed to try and recover from such cases and will attempt to resubmit reports up to 5 times. The message you are displaying is a little bit odd, because I can't reproduce it in my testing. Whenever I attempt to reproduce your error case, by simulating a connection error, I see something like:
```
2022-09-29T08:32:05.348547Z  WARN apollo_router::spaceport::server: attempt: 5, could not transfer: error sending request for url (https://usage-reporting.api.apollographql.com/api/ingress/traces): error trying to connect: tcp connect error: Connection refused (os error 61)
2022-09-29T08:32:05.600182Z ERROR apollo_router::spaceport::server: report submission failed: status: Unavailable, message: "error sending request for url (https://usage-reporting.api.apollographql.com/api/ingress/traces): error trying to connect: tcp connect error: Connection refused (os error 61)", details: [], metadata: MetadataMap { headers: {} }
```
(truncated to remove the first 4 attempts to transfer the report...)
This error is actually occurring within spaceport when it tries to talk to the usage-reporting endpoint, whereas your error is occurring much sooner in the transfer process within the router as data is transferred to spaceport and I can't actual re-create this in my test environment (Mac OS).
@hrkfdn 
Some questions:
1. Can you describe the environment your router is running in:
 - Is it containerised (docker, k8s, etc..)?
 - Is it Linux/MacOS/Windows?
 - If containerised, are there any resource constraints?
2. Could you clarify your comment about "schema updates"? Issues relating to transferring statistics to spaceport should not affect schema update behaviour.
3. Can you check your logs for additional error messages? In particular, any at level ERROR.

**Comment by hrkfdn** [comment]
Hey @garypen, thanks for your response.

I agree it seems odd. It hasn't recovered since I opened the ticket. At first I suspected socket limit exhaustion, but that would be a different error, no? In that case cURLing the Spaceport URL probably also wouldn't work. At the same time we are not encountering any problems serving client requests.

1. The router runs in K8s on Linux. I checked memory usage and it seems fine. The open file ulimit is set to 1048576, procfs lists 65 open file descriptors. Here's the output of `ulimit -a`:
  ```
  real-time non-blocking time  (microseconds, -R) unlimited
  core file size              (blocks, -c) unlimited
  data seg size               (kbytes, -d) unlimited
  scheduling priority                 (-e) 0
  file size                   (blocks, -f) unlimited
  pending signals                     (-i) 125164
  max locked memory           (kbytes, -l) 64
  max memory size             (kbytes, -m) unlimited
  open files                          (-n) 1048576
  pipe size                (512 bytes, -p) 8
  POSIX message queues         (bytes, -q) 819200
  real-time priority                  (-r) 0
  stack size                  (kbytes, -s) 8192
  cpu time                   (seconds, -t) unlimited
  max user processes                  (-u) unlimited
  virtual memory              (kbytes, -v) unlimited
  file locks                          (-x) unlimited
  ```
2. We use Apollo Studio schema checks to detect breaking changes in our schema: https://www.apollographql.com/docs/studio/schema-checks/. I assumed this relies on usage statistics and lack of these statistics would render the checks useless, but it's just an assumption.
3. No errors besides this (unfortunately?).

**Comment by garypen** [comment]
Hi @hrkfdn thanks for the extra information. I artificially limited the number of file descriptors and saw errors like:
```
2022-09-29T11:45:55.670454Z ERROR apollo_router::services::subgraph_service: fetch_error="hyper::Error(Connect, ConnectError(\"tcp open error\", Os { code: 24, kind: Uncategorized, message: \"Too many open files\" }))"
2022-09-29T11:45:55.670504Z  INFO apollo_router::plugins::include_subgraph_errors: redacted subgraph(accounts) error
2022-09-29T11:45:55.670525Z ERROR apollo_router::query_planner::execution: Fetch error: HTTP fetch failed from 'accounts': HTTP fetch failed from 'redacted': redacted
2022-09-29T11:45:55.670698Z ERROR apollo_router::services::subgraph_service: fetch_error="hyper::Error(Connect, ConnectError(\"tcp open error\", Os { code: 24, kind: Uncategorized, message: \"Too many open files\" }))"
2022-09-29T11:45:55.670716Z ERROR apollo_router::services::subgraph_service: fetch_error="hyper::Error(Connect, ConnectError(\"dns error\", Custom { kind: Uncategorized, error: \"failed to lookup address information: nodename nor servname provided, or not known\" }))"
```
so, I don't think that's the cause of your problem.

regarding 2. If you don't have any statistics, it would impact breaking change detection. Thank you for clarifying.

I'll try to reproduce the problem in a test cluster and will continue to investigate. In the meantime, if you haven't already, I recommend a rolling restart on your router deployment to bounce your pods and restore functionality to the studio reporting.

**Comment by hrkfdn** [comment]
Thanks @garypen, if you need any more information please let me know.

**Comment by hrkfdn** [comment]
Hi @garypen, I have some more observations that could be relevant:

```
apollo-server-5b9645cb47-bffzk apollo-server {"timestamp":"2022-10-04T10:34:03.533146Z","level":"INFO","fields":{"message":"reloading schema"},"target":"apollo_router::state_machine"}
apollo-server-5b9645cb47-bffzk apollo-server {"timestamp":"2022-10-04T10:34:05.501875Z","level":"INFO","fields":{"message":"healthcheck endpoint exposed at http://0.0.0.0:8088/health"},"target":"apollo_router::axum_factory::axum_http_server_factory"}
apollo-server-5b9645cb47-bffzk apollo-server {"timestamp":"2022-10-04T10:34:05.501978Z","level":"INFO","fields":{"message":"GraphQL endpoint exposed at http://0.0.0.0:4000/graphql 🚀"},"target":"apollo_router::axum_factory::axum_http_server_factory"}
apollo-server-5b9645cb47-bffzk apollo-server {"timestamp":"2022-10-04T10:34:05.943225Z","level":"INFO","fields":{"message":"terminating apollo exporter"},"target":"apollo_router::plugins::telemetry::apollo_exporter"}
apollo-server-5b9645cb47-bffzk apollo-server {"timestamp":"2022-10-04T10:34:08.461725Z","level":"WARN","fields":{"message":"failed to submit stats to spaceport: status: Unavailable, message: \"error trying to connect: tcp connect error: Connection refused (os error 111)\", details: [], metadata: MetadataMap { headers: {} }"},"target":"apollo_router::plugins::telemetry::apollo_exporter"}
apollo-server-5b9645cb47-bffzk apollo-server {"timestamp":"2022-10-04T10:34:14.461693Z","level":"WARN","fields":{"message":"failed to submit stats to spaceport: status: Unavailable, message: \"error trying to connect: tcp connect error: Connection refused (os error 111)\", details: [], metadata: MetadataMap { headers: {} }"},"target":"apollo_router::plugins::telemetry::apollo_exporter"}
...
```
It seems to happen when the schema is pulled from Apollo Studio. Interestingly this happened across all 3 pods at roughly the same time (± a few seconds).

**Comment by garypen** [comment]
@hrkfdn  Ok. That's interesting info. I think that what is happening is that back-end of the connection (inside the router) is going away when the schema is reloaded.

My mitigation in https://github.com/apollographql/router/pull/1928 will certainly help with this problem since it will cause the "client" (internal to the router) to re-connect to the backend (also internal to the router) and should pick up the new back-end and be able to continue submitting stats. I'll try to verify that based on the schema reload information for testing.

**Comment by garypen** [comment]
@hrkfdn Another question: Do you have hot reload (--hot-reload) enabled?

**Comment by hrkfdn** [comment]
> @hrkfdn Another question: Do you have hot reload (--hot-reload) enabled?

Unless it's enabled by default, nope.

**Comment by garypen** [comment]
@hrkfdn I hope that your problems are resolved in today's release: 1.2.0. Thanks a lot for your input, it was very helpful when I was trying to narrow down the source of the problem.

**Comment by hrkfdn** [comment]
Hey @garypen, metrics have been stable for a few days now, so it looks good. Thanks for tackling this + glad my input helped.
